[클린코드 리액트 2기 장태영] 페이먼츠 미션 Step 2

taeyoungs opened the pull request:
- [데모 페이지](https://react-payments-git-step2-taeyoungs.vercel.app/)
- [스토리북](https://63f1c654e4f702113aa32a89-pnzdnybhol.chromatic.com/?path=/story/common-button--template)

## 필수 요구사항
- [x] Storybook 상호 작용 테스트
- [ ] Controlled & Uncontrolled Components에 입각하여 Form 핸들링
- [x] Context API를 활용해 전역 상태 관리 및 계층 재구성

## 카드 추가 확인

- [x] 이전 폼에서 입력된 카드를 보여준다.
- [x] 카드 별칭을 입력할 수 있다.
   - [x] placeholder는 카드 별칭 (선택)이다.
   - [x] 빈 입력값인 경우, 카드사 이름이 별칭으로 저장된다.
   - [x] 최대 길이는 10자리이다.
- [x] 확인 버튼을 누르면, 카드 목록 페이지로 이동한다.

## 카드 목록

- [x] 카드 목록을 조회할 수 있다.
- [x] 카드 목록은 최신순(내림차순)으로 정렬된다.
- [x] 목록 최상단에 +을 누르면 카드 추가 페이지로 이동한다.
- [x] 카드를 클릭하면, 카드 별칭 수정(카드 추가 완료 페이지)로 이동한다.
- [x] 카드를 삭제할 수 있다.

### 기타

제 부주의로 rebase를 진행하지 않은 채 브랜치를 생성하고 커밋을 진행해서 제대로 PR을 올린건지 의문입니다. 🥲
(슬랙에서 준영님이 언급하셨던 충돌 상황 같습니다)
NEXT STEP 레포에 있는 [문서](https://github.com/next-step/nextstep-docs/blob/master/codereview/review-step3.md)대로 충돌을 해결하긴 했는데 PR에 여전히 문제가 있다면 고치도록 하겠습니다!

taeyoungs committed:
init: Project initialize (CRA + storybook)

taeyoungs committed:
chore: styles 폴더 이동 및 index.css import

taeyoungs committed:
feat: router 객체 및 Layout 컴포넌트 추가

taeyoungs committed:
feat: Card, CardList 컴포넌트 템플릿 추가

taeyoungs committed:
feat: 카드 추가 페이지 템플릿 추가

taeyoungs committed:
mod: Card 컴포넌트 업데이트

taeyoungs committed:
chore: 불필요한 컴포넌트 삭제

taeyoungs committed:
chore: 네이밍 수정 (serial => card)

taeyoungs committed:
feat: Input 컴포넌트 구현 및 헬퍼 타입 추가

taeyoungs committed:
feat: 카드 번호 관리를 위한 커스텀 훅 추가

taeyoungs committed:
feat: 카드 번호 컨테이너 컴포넌트 구현

taeyoungs committed:
mod: Input 컴포넌트 forwardRef 추가

taeyoungs committed:
feat: Dash 컴포넌트 구현

components/Dash/index.tsx
- Card 컴포넌트와 CardNumberContainer 컴포넌트에서 사용할
  '-'를 표현할 Dash 컴포넌트 구현

taeyoungs committed:
feat: Input focus 컨트롤을 위한 커스텀 훅 추가

taeyoungs committed:
mod: ref와 useInputFocusing을 이용한 focus 컨트롤

taeyoungs committed:
feat: Card 뒷 8자리 마스킹 처리 추가

taeyoungs committed:
mod: isNumber 함수를 유틸 함수로 분리

taeyoungs committed:
feat: 만료일 input 관련 커스텀 훅 구현

taeyoungs committed:
feat: 만료일 컴포넌트 ExpiredDateContainer 구현

taeyoungs committed:
mod: input placeholder 수정

taeyoungs committed:
mod: Card 컴포넌트 expiredDate 매핑

taeyoungs committed:
feat: InputProgress 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
chore: 불필요한 스토리북 파일 삭제

taeyoungs committed:
feat: CompanyColor 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
feat: Masking 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
mod: Masking 컴포넌트 CSS 업데이트

taeyoungs committed:
mod: '*' 문자열을 Masking 컴포넌트로 대체

taeyoungs committed:
mod: Input 컴포넌트 align 속성 추가

taeyoungs committed:
chore: components 폴더의 하위 폴더 구조 변경

taeyoungs committed:
chore: 공용 컴포넌트 스토리북 파일 위치 변경

taeyoungs committed:
feat: Dash 컴포넌트 스토리북 추가

taeyoungs committed:
feat: CardOwnerContainer 및 useCardOwner 구현

taeyoungs committed:
mod: Input, Card 컴포넌트 CSS 업데이트

taeyoungs committed:
mod: Input 컴포넌트 CSS 업데이트

taeyoungs committed:
mod: Input size 속성 추가 및 스토리북 수정

taeyoungs committed:
feat: SecretCodeContainer(보안코드) 컴포넌트 구현

taeyoungs committed:
mod: Masking 컴포넌트 CSS 업데이트

taeyoungs committed:
feat: CardPasswordContainer(비밀번호) 컴포넌트 구현

taeyoungs committed:
feat: CardCompanyPicker(카드사 선택) 컴포넌트 구현

taeyoungs committed:
chore: CompanyColor, Card 컴포넌트 CSS 업데이트

taeyoungs committed:
mod: 카드 추가 페이지 CardCompanyPicker 추가

taeyoungs committed:
feat: 카드 등록 확인 페이지 템플릿 추가

taeyoungs committed:
chore: 불필요한 스토리북 파일 삭제

taeyoungs committed:
feat: Button 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
feat: Header 컴포넌트 구현 및 스토리북 추가

taeyoungs committed:
feat: 페이지 이동 버튼 추가

taeyoungs committed:
chore: 사용되지 않는 masking 함수 삭제

taeyoungs committed:
Merge branch 'step1'

taeyoungs committed:
chore: chromatic deploy

taeyoungs committed:
Merge branch 'step1'

taeyoungs committed:
chore: chromatic deploy

taeyoungs committed:
mod: router 객체 분리 및 라우팅 경로 상수화

taeyoungs committed:
chore: typescript를 devDependencies로 이동

taeyoungs committed:
chore: EOL 제거

taeyoungs committed:
chore: Masking, CompanyColor를 도메인 폴더 하위로

taeyoungs committed:
chore: import 경로 오류 수정

taeyoungs committed:
mod: Dash 컴포넌트 업데이트

components/common/Dash
- visible 속성 삭제 후 외부에서 Dash 컴포넌트를 조건부 렌더링하는
  방식으로 변경
- width 속성을 받아 padding에 바인딩해주는 상태가 혼란을 줄 수도
  있다고 생각하여 after 선택자를 이용한 방식으로 변경

taeyoungs committed:
mod: Card 컴포넌트 업데이트

componsnts/domain/Card
AddingCard/hooks/useCompanyPicker.ts
- company 상태값에 대한 초기값을 부여하고 null을 제거하여
  Card 컴포넌트 내에서 null 값으로 인한 삼항 연산자를 제거

taeyoungs committed:
chore: pixelOf 유틸 함수 분리

taeyoungs committed:
mod: 삼항 연산자를 pixelOf 함수로 대체

taeyoungs committed:
mod: 카드 추가 페이지 내 input name 문자열 상수화

taeyoungs committed:
mod: Input 컴포넌트 업데이트

taeyoungs committed:
chore: .root에 대한 CSS 업데이트

taeyoungs committed:
mod: Label 컴포넌트 활용

taeyoungs committed:
feat: 여러 개의 Ref를 생성해서 반환하는 훅 추가

taeyoungs committed:
mod: 보안코드 상태값도 커스텀 훅으로 분리

taeyoungs committed:
mod: Object.values(CardCompany) 상수 처리

taeyoungs committed:
mod: isOutOfRangeMonth 함수를 유틸 함수로 분리

taeyoungs committed:
chore: 오타 수정

taeyoungs committed:
mod: useInputFocusing 내 중복 콜백 함수 분리

taeyoungs committed:
chore: chromatic deploy

taeyoungs committed:
Merge branch 'step1'

taeyoungs committed:
mod: Header 컴포넌트 prefix 속성 네이밍 변경

taeyoungs committed:
chore: 오타 수정

taeyoungs committed:
Merge branch 'step1'

taeyoungs committed:
mod: CardBaseProps 타입 추가

types/card.ts
domain/Card/index.ts
- Card 컴포넌트의 props 타입을 재사용하기 위해 CardBaseProps라는
  인터페이스로 분리

AddingCard/index.tsx
- 카드 완료 페이지로 넘어갈 때 form 하위에 존재하는 제어 컴포넌트
  값들을 state 옵션으로 들고 가도록 navigate state 추가

taeyoungs committed:
mod: ConfirmCard 컴포넌트 업데이트

contants/card.ts
- 카드 별칭 최대 자리 수 상수값 추가

ConfirmCard/index.tsx
- 카드 별칭를 컨트롤하는 제어 컴포넌트 추가

taeyoungs committed:
feat: 카드 목록을 관리하는 Context 추가

CardContextProvider/index.tsx
- 보유 카드 목록을 value로 갖고 있는 Context 생성
- 카드 추가와 삭제에 대한 Action 및 reducer 생성
- dispatch만 사용하는 컴포넌트가 있을 수도 있고 state만 사용하는
  컴포넌트가 있을 수 있기 때문에 dispatch와 state에 대한 Context
  를 각각 생성

CardContextProvider/model/index.ts
- CardContextProvider 내에서 존재하는 reducer에서 사용될 로직을
  분리하여 model이라는 키워드 아래에 정리

taeyoungs committed:
feat: CardContextProvider를 위한 커스텀 훅 추가

taeyoungs committed:
feat: Card에 대한 인터페이스 추가

taeyoungs committed:
mod: Action 네이밍 수정 (add => addOrUpdate)

taeyoungs committed:
mod: CardContext의 value로 카드 목록 렌더링

taeyoungs committed:
chore: Card 타입 네이밍 수정 (Card => CardType)

taeyoungs committed:
chore: .root 클래스 및 a 태그 css 업데이트

taeyoungs committed:
mod: 카드 별칭 유무에 따른 타이틀 변경

taeyoungs committed:
feat: HoldingCard 컴포넌트 추가

taeyoungs committed:
chore: CardType 네이밍 변경 (CardType => ICard)

taeyoungs committed:
chore: 스토리북 카테고리 추가 (domain, common)

taeyoungs committed:
feat: Card 컴포넌트에 대한 스토리북 추가

taeyoungs committed:
mod: input required 옵션 추가

taeyoungs committed:
mod: input required 옵션 추가 및 .app CSS 업데이트

taeyoungs committed:
mod: useRouter 훅에 이전 페이지 이동 함수 추가

taeyoungs committed:
chore: chromatic deploy

taeyoungs committed:
Merge branch 'step2'

taeyoungs committed:
fix: Conflicts 해결

taeyoungs committed:
feat: 빈 입력값인 경우, 카드사 이름이 별칭으로 저장

InSeong-So commented on src/components/common/Button/Button.stories.tsx:
```diff
@@ -4,7 +4,7 @@ import { Button } from 'components/common';
 import { ButtonProps } from 'components/common/Button';
 
 export default {
-  title: 'Button',
+  title: 'Common/Button',
   component: Button,
```
잘 묶어주셨네요!! 이런 부분은 사소하면서도 DX를 크게 향상시키기 때문에 세심하게 작업하는 게 좋습니다👍

InSeong-So commented on src/components/domain/Card/index.tsx:
```diff
@@ -6,17 +6,16 @@ import { Masking } from 'components/domain';
 
-interface CardProps {
+export interface CardProps {
   company?: CardCompany;
-  name: string;
+  cardOwner: string;
   cardNumber: CardNumber;
   expiredDate: ExpiredDate;
 }
```
이 타입은 상당히 전역적으로 쓰일 것 같은데, types 파일에 분리하거나 import 해오는 것이 좋을 것 같습니다😊

일반적으로 컴포넌트와 타입을 같이 export하면 트리쉐이킹 할 때 정규식 작성하기가 상당히 까다로워지거든요. 혹은 동일 파일임에도 import 경로를 다르게 해야하는 번거로움이 있답니다!

taeyoungs replied:
> 일반적으로 컴포넌트와 타입을 같이 export하면 트리쉐이킹 할 때 정규식 작성하기가 상당히 까다로워지거든요. 혹은 동일 파일임에도 import 경로를 다르게 해야하는 번거로움이 있답니다!

넵넵 알겠습니다! 생각해보니 현재 형태는 `CardProps`만을 다른 곳에서 import 하는 상황에서도 바람직한 모양새가 아닐 것 같네요! 👍

InSeong-So commented on src/components/domain/HoldingCard/index.tsx:
```diff
@@ -0,0 +1,48 @@
+  const dispatch = useCardDispatch();
+
+  const handleClickDeleteButton = () => {
+    const isCanDeleted = window.confirm(CONFIRM_MESSAGE);
+
+    if (isCanDeleted) {
+      dispatch({ type: 'DELETE_CARD', payload: card });
+    }
```
confirm이 주는 `yes or no`는 상당히 명확하므로 인라인으로 작성해주는 것도 좋습니다🤗

```suggestion
    if (window.confirm(CONFIRM_MESSAGE)) {
      dispatch({ type: 'DELETE_CARD', payload: card });
    }
```

그리고 여러 컴포넌트, 페이지 단위에서 dispatch를 호출하는 것보다 커스텀 훅 또는 핸들러를 만들어두고 인자만 넘겨주는 형태로 작성해주는 것이 현재 태영님이 짠 구조에서 좀 더 직관적일 것 같아요.

```ts
// 어딘가에 정의한... 저는 context를 선언한 폴더 내에서 co-location을 염두에 두면 좋겠네요!
export const deleteCard = (card: ICard) => {
  dispatch({ type: 'DELETE_CARD', payload: card });
};

// src/components/domain/HoldingCard/index.tsx
import { deleteCard } from '..//..';

// 생략
    if (window.confirm(CONFIRM_MESSAGE)) deleteCard(card);
```

taeyoungs replied:
Redux 사용할 때 action을 만드는 형태와 비슷한 것 같네요! 제안해주신 형태가 해당 함수가 무엇을 하려고 하는지 대략적으로 알기에 더 적합한 것 같습니다. 

인성님 말씀처럼 co-location을 염두해두고 action을 모아보도록 하겠습니다. :)

InSeong-So commented on src/types/card.ts:
```diff
@@ -25,3 +25,11 @@ export type CardPassword = {
 };
+
+export interface ICard {
+  cardOwner: string;
+  cardNumber: CardNumber;
+  expiredDate: ExpiredDate;
+  company: CardCompany;
+  alias?: string;
```
조오오오금 네이밍을 통합해도 괜찮을 것 같아요! card 접두어를 넣을건지... 뺄것인지...!

taeyoungs replied:
~~앗 .. 넣는 쪽으로 통일하겠습니다!~~

수정을 하면서 느낀건데 접두어를 넣어서 진행하게 되면 `card`라는 객체에서 특정 속성을 참조할 때 접두어가 불필요하게 한번 더 붙는 느낌이 있더라구요. 그래서 빼는 쪽으로 선회했습니다! 🫠

InSeong-So commented on src/pages/HoldingCards/index.tsx:
```diff
@@ -1,45 +1,29 @@
+      <Link to="new" className="card-box">
         <div className="empty-card">+</div>
       </Link>
+      {cards.map((card) => (
+        <HoldingCard key={cardNumberToString(card)} card={card} />
+      ))}
     </div>
   );
 }
 
+const cardNumberToString = (card: ICard) => Object.values(card.cardNumber).toString();
```
이게 뭘까 골똘히 생각했는데😅 `cardNumberToString` 함수가 필요했을까 싶네요...! card.cardNumber로 정의하면 안되었을까요?

taeyoungs replied:
카드 번호만이 유니크한 값이 될 것이라 생각해서 카드 번호들을 하나의 문자열로 만들어서 전달해준 것이긴 한데 .. 이럴 거면 카드 객체에 유니크 아이디를 포함시켰어야 했을까요 ?? 🥲

InSeong-So commented on src/hooks/useRouter.ts:
```diff
@@ -0,0 +1,20 @@
+import { useLocation, useNavigate } from 'react-router-dom';
+
+const useRouter = <T>() => {
+  const navigate = useNavigate();
+  const location = useLocation();
+
+  const locationState = location.state as T;
+
+  const goBack = () => {
+    navigate(-1);
+  };
+
+  return {
+    navigate,
+    locationState,
+    goBack,
+  };
+};
```
이 코드는 조금 더 고도화시킬 수 있어보입니다🤗 훅으로 감싼 시도는 매우 좋았는데, 사실 변한게 크게 없어요. goBack 정도... 어떤 라우터가 있는지 상세하게 선언해두거나, 히스토리를 관리한다거나... 여러 방안을 적용하다보면 정말 괜찮은 커스텀 훅을 만들 수 있답니다😊

taeyoungs replied:
> 이 코드는 조금 더 고도화시킬 수 있어보입니다🤗 훅으로 감싼 시도는 매우 좋았는데, 사실 변한게 크게 없어요.

공감합니다! Step 1에서 인성님 피드백이 떠올라서 훅으로 감싸긴 했는데 goBack을 제외하면 크게 의미가 없는 상태라고 느끼긴 했습니다..
히스토리 관리는 당장 생각나는 좋은 방법이 없어서 "어떤 라우터가 있는지 상세하게 선언해두거나" 요 방향으로 조금씩 개선해보도록 하겠습니다. 🙇‍♂️

InSeong-So commented on src/contexts/CardContextProvider/hooks/useCardDispatch.ts:
```diff
@@ -0,0 +1,11 @@
+import { useContext } from 'react';
+
+import { CardDispatchContext } from 'contexts/CardContextProvider';
+
+const useCardDispatch = () => {
+  const dispatch = useContext(CardDispatchContext);
+
+  return dispatch;
+};
+
+export default useCardDispatch;
```
`useCardState`와 같이 있어도 괜찮지 않을까요~? 분리할 수준의 추상화나 사용처가 아닌 것 같아요!

taeyoungs replied:
넵넵 불필요한 추상화 같네요! `useCardState` 쪽에 편입시키도록 하겠습니다.

InSeong-So commented on src/pages/AddingCard/index.tsx:
```diff
@@ -23,20 +25,27 @@ function AddingCard() {
+      cardOwner,
+      company,
+    };
+
+    navigate(PATHS.CONFIRM, { state: navigateState });
   };
```
1. navigate를 그대로 export하면서 상수 주입
2. navigate를 추상화하여 url을 타입 리터럴로 선언하여 관리하기
3. 1+2를 적절히 혼합하여 사용하기

취사 선택하면 좋을 것 같아요😄

taeyoungs replied:
useRouter라는 훅을 만들었으니 2번 쪽으로 통일해서 써보도록 하겠습니다.
경로에 대한 상수도 해당 훅과 같이 있으면 개인적으로 관리 측면에도 도움이 될 것 같아서요!

InSeong-So commented on src/pages/ConfirmCard/index.tsx:
```diff
@@ -1,37 +1,58 @@
+  };
+
+  const handleClickNextButton = () => {
+    const newAlias = alias === '' ? card.company : alias;
+
```
이 구문도 title과 같은 스코프에 두시고, title과 newAlias는 useMemo를 사용하는 건 어떨까요?

taeyoungs replied:
오오 좋은 것 같습니다. 바로 적용해보도록 하겠습니다. 🚀

InSeong-So commented on src/contexts/CardContextProvider/model/index.ts:
```diff
@@ -0,0 +1,23 @@
+};
+
+export const addOrUpdateCard = (cards: ICard[], newCard: ICard) => {
+  const existedCard = cards.find((card) => isMatchedCard(card, newCard));
+
+  if (!existedCard) {
+    return [newCard, ...cards];
+  }
+
+  const updatedCards = cards.map((card) => (isMatchedCard(card, newCard) ? newCard : card));
```
find, map이 똑같은 타겟으로 비슷하게 로직을 실행하는데... reduce 등을 이용해서 처리하는 것은 어떨까요?

taeyoungs replied:
reduce를 활용할 생각을 못했네요!
한번 적용해보도록 하겠습니다. 🚀

InSeong-So reviewed (changes requested):
태영님, 안녕하세요! 리뷰어 소인성입니다🙇‍♂️
미션을 잘 수행해주셨네요! 코멘트가 많지 않습니다😄

기능 에러만 좀 보고 갈게요!
1. MM, YY에 0 입력 가능
2. MM, YY가 입력되지 않아도 `/` 잔재
![image](https://user-images.githubusercontent.com/18283006/222875880-56c59b9e-546e-4e3d-a8ad-e5af355ca8bf.png)

코멘트만 확인해주세요!
머지해도 괜찮을 것 같으나 한 번 정도는 보시고 싶지 않을까하여 재요청드립니다☺

taeyoungs committed:
chore: ExpiredDate 타입을 CardExpiredDate로 변경

taeyoungs committed:
chore: 컴포넌트를 제외한 cardOwner에서 접두어 제거

taeyoungs committed:
chore: 보안코드 래핑 컴포넌트에 접두어 추가

taeyoungs committed:
chore: 컴포넌트를 제외한 cardNumber에서 접두어 제거

taeyoungs committed:
chore: ExpiredDateContainer에 'Card' 접두어 추가

taeyoungs committed:
mod: useCardState와 useCardDispatch 통합

taeyoungs committed:
feat: 카드 삭제, 추가 및 업데이트 함수 분리

useCardActions.ts
- 컴포넌트에 dispatch를 노출시키는 방식 대신 노출시켜서 하려던
  로직을 개별 함수로 담고 있고 해당 함수들을 반환하는 커스텀 훅을
  하나 추가하여 이를 사용하는 방식으로 변경

taeyoungs committed:
mod: title과 newAlias 변수의 depth 통일 및 useMemo

taeyoungs committed:
chore: 오타 수정

taeyoungs committed:
mod: CardProps를 ICard 타입을 이용한 방식으로 변경

taeyoungs committed:
mod: ICard에 id 속성 및 ICardWithoutId 타입 추가

types/card.ts
- 카드 목록을 업데이트하거나 삭제할 때 이용할 id 속성을 ICard
  타입에 추가
- 새로운 카드가 카드 목록에 추가되기 전 상태를 위한, 즉 카드에
  고유 id 속성이 생기기 전 상황을 위한 ICardWithoutId 타입 추가

taeyoungs committed:
mod: 통합되어 있던 카드 추가와 업데이트 로직을 분리

CardContextProvider/
- addOrUpdateCard의 네이밍으로 새로운 카드를 추가하고 기존 카드를
  업데이트하던 함수를 addCard와 updateCard라는 네이밍의 두 가지
  함수로 분리
- 카드가 추가될 때마다 고유의 id 값을 생성
- id를 이용하여 카드를 삭제 또는 업데이트하도록 변경

taeyoungs committed:
mod: 변경된 카드 삭제, 추가, 업데이트 방식으로 전환

taeyoungs committed:
mod: 불필요한 함수 삭제 및 key 속성 변경

taeyoungs committed:
chore: theme 객체 네이밍을 Uppercase로 변경

taeyoungs committed:
mod: navigate 함수를 래핑한 go, replace 함수 추가

useRouter.ts
- Router 객체에 없는 경로를 navigate 함수에 전달하는 것을 방지하기
  위해 간단하게 navigate를 래핑한 함수를 만들어 'to' 파라미터에
  올 수 있는 경로값을 제한

taeyoungs committed:
feat: 번호, 만료일, 보안코드 최소 길이 옵션 추가

taeyoungs committed:
chore: Card 컴포넌트 내 만료일 CSS 업데이트

taeyoungs committed:
mod: 현재 년도 이하 년도는 입력하지 못하게 변경

utils.ts
- 현재 년도 뒤 두자리를 구해 파라미터로 넘어오는 년도가 현재 년도
  보다 작을 경우 false를 반환하는 헬퍼 함수 추가

taeyoungs committed:
chore: ExpiredDateContainer 네이밍에 접두어 추가

taeyoungs commented:
안녕하세요 인성님 !

> 머지해도 괜찮을 것 같으나 한 번 정도는 보시고 싶지 않을까하여 재요청드립니다 ☺

감사합니다! 기능 에러 부분은 생각치도 못했던 부분(기능 에러 외의 다른 부분도)이라 짚어주셔서 감사합니다. 🙇‍♂️

## 피드백 체크리스트

- [x]  CardProps를 컴포넌트에서 export 하지 말고 분리하기
- [ ]  useRouter 고도화
    - [x]  navigate를 추상화하여 url을 타입 리터럴로 선언하여 관리 및 이를 재사용하기
- [x]  `card` 접두어를 일관적으로 적용하기
- [x]  co-location을 염두해두고 dispatch를 래핑한 action을 만들어서 재사용하기
- [x]  useCardDispatch를 useCardState에 편입시키기
- [x]  `newAlias = alias === '' ? card.company : alias` 구문을 title과 같은 스코프에 두고, 
title과 newAlias는 useMemo를 사용해보기
- [x]  만료일에 0이 입력되는 경우 방지
- [x]  만료일에 0만 입력해서 등록이 되는 경우 방지

## 체크리스트 세부사항

### useRouter 고도화

> 이 코드는 조금 더 고도화시킬 수 있어보입니다🤗 훅으로 감싼 시도는 매우 좋았는데, 사실 변한게 크게 없어요. goBack 정도... 어떤 라우터가 있는지 상세하게 선언해두거나, 히스토리를 관리한다거나... 여러 방안을 적용하다보면 정말 괜찮은 커스텀 훅을 만들 수 있답니다😊

엄청난 변화를 주지는 못한 것 같습니다.  🫠 현재 변화를 준건 `navigate` 함수를 래핑해서 래핑한 함수를 사용할 때 미리 정의해둔 경로가 아니라면 타입 에러를 발생시키도록 만든 정도입니다.

### `card` 접두어

쓰읍 .. 이게 기준을 두고 정리하긴 했는데 .. 더 혼란이 가중된 건 아닌가 .. 싶네요. 🥲
- `card` 객체의 속성 중에 `card`라는 접두어를 가지고 있는 속성들은 접두어를 제거했습니다. 아무래도 `card.cardNumber`와 같은 식으로 객체의 속성을 접근하는 게 뭔가 괴리감이 들었습니다.
- 컴포넌트와 타입에는 접두어를 유지한 상태입니다.

이렇다 보니 결국 다음과 같이 수정이 된 상태입니다. 

```typescript
// #1 타입
export interface ICard {
  id: number;
  owner: string;
  numbers: CardNumber;
  expiredDate: CardExpiredDate;
  company: CardCompany;
  alias?: string;
}

// #2 컴포넌트
<CardExpiredDateContainer />

// #3 객체
card.expiredDate
```

### 만료일 관련

- 인성님이 짚어주신 만료인 외에도 각 카드 번호를 4자리 채우지 않고도 폼 양식을 제출할 수 있는 상황이었습니다. `input` 태그에 `minLength` 속성을 이용해서 이런 케이스를 방지하게 만들었습니다.
- `isOutOfRangeYear`라는 함수를 추가하여 현재 년도 이하(ex. 현재 2023년인데 만료일을 22년도로 하려는 경우)로는 입력을 못하게 만들었습니다.
- 만료일 중 '월'은 첫 번째 자리 수까지는 0을 입력 가능하게 만들었습니다.

지금 전체적으로 `input` 태그가 가진 기본 속성들을 이용해 validation을 처리하거나 요구 사항에 어긋나는 범위의 입력은 막고 있는 상황인데 아예 입력을 막다 보니까 사용성이 떨어지는 감이 있는 것 같습니다. 입력을 막기보다는 다르게 입력해야 한다는 것을 폼 양식을 제출할 때 각 `input` 근처에서 표시해줘야하나 .. 둘 다 해주는게 맞는걸까 .. 고민이 되네요. 😇

### `addOrUpdateCard` 함수 변경

> find, map이 똑같은 타겟으로 비슷하게 로직을 실행하는데... reduce 등을 이용해서 처리하는 것은 어떨까요?

코멘트는 위와 같이 남겨주시긴 했는데 이미 하나의 함수가 2가지 일(배열을 전체 순회해서 논리값을 만들어 내는 일과 배열 내 특정 값을 업데이트하는 일)을 해서 그런지 이를 reduce로 통합하려니 코드가 점점 알아보기 힘들어 지는 상황이 발생했습니다(플래그 변수를 만들거나 reduce 함수가 무얼 하려는 건지 알 수 없게 로직을 작성하고 있는 상황을 맞이했었습니다. 😂).

그래서 선택한 방법은 카드 객체에 고유 id값을 추가하고 추가하는 로직과 업데이트하는 로직을 분리해서 2개의 함수로 만드는 것이었습니다. 

카드 번호를 하나의 문자열로 만든 것을 고유 id로 사용하려다 카드 번호가 겹치는 상황을 체크하는 로직을 구현하지 않은 상황이라 현재는 고유 id는 `new Date().getTime()`을 이용해서 만들고 있습니다.

InSeong-So commented on src/components/domain/Card/index.tsx:
```diff
@@ -4,20 +4,16 @@ import { Dash } from 'components/common';
+import { CardCompany, ICardWithoutId } from 'types/card';
 
-interface CardProps {
+export interface CardProps extends Omit<ICardWithoutId, 'company'> {
   company?: CardCompany;
```
👏👏👏 저도 이렇게 interface를 활용하는 편입니다.
여기서 조금 더 확장해보자면, ICardWithoutId은 도메인에 집중된 Props니까 컴포넌트 레벨의 타입 Props와 도메인 Props를 두 개 준비해서 extends로 상속 받으면 어떨까요?

taeyoungs replied:
> 여기서 조금 더 확장해보자면, ICardWithoutId은 도메인에 집중된 Props니까 컴포넌트 레벨의 타입 Props와 도메인 Props를 두 개 준비해서 extends로 상속 받으면 어떨까요?

오 .. 이런 식으로 구분해서 생각해본 적은 없는 것 같아요. 🤔
Step 3 요구사항 구현 전에 정리해서 적용해보도록 하겠습니다!

InSeong-So replied:
@taeyoungs 예를 들어보자면... 가령 input 엘리먼트에 추가적으로 주입해야 하는 요소가 존재하는 경우...

```ts
// 컴포넌트 props
type InputElementProps = InputHTMLAttributes<HTMLInputElement>;

// 비즈니스 props
type CardOptions = {
  id: number;
  owner: string;
  numbers: CardNumber;
  expiredDate: CardExpiredDate;
  company: CardCompany;
  alias?: string;
}

// 카드 컴포넌트 props
interface CardProps extends InputElementProps, Partial<CardOptions> {}
```

이렇게를 이야기한 것이니, 참고해주세요🤗

taeyoungs replied:
아 ~ 이해했습니다!
예시까지 들어서 설명해주셔서 감사합니다. 🙇‍♂️
두 가지가 섞여 있어서 불편하거나 헷갈리는 경우가 있곤 했는데 이런 식으로 구분 지어주면 상당히 좋겠네요!

InSeong-So commented on src/contexts/CardContextProvider/hooks/useCardActions.ts:
```diff
@@ -0,0 +1,25 @@
+const useCardActions = () => {
+  const dispatch = useContext(CardDispatchContext);
+
+  const addCard = (card: ICardWithoutId) => {
+    dispatch({ type: ACTION_TYPE.ADD_CARD, payload: card });
+  };
+
+  const updateCard = (card: ICard) => {
+    dispatch({ type: ACTION_TYPE.UPDATE_CARD, payload: card });
+  };
+
+  const deleteCard = (id: number) => {
+    dispatch({ type: ACTION_TYPE.DELETE_CARD, payload: id });
+  };
```
👍

InSeong-So commented on src/pages/AddingCard/hooks/useNumbers.ts:
```diff
@@ -0,0 +1,29 @@
+const useNumbers = () => {
+  const [numbers, setNumbers] = useState(INITIAL_STATE);
+
+  const handleChangeNumbers: React.ChangeEventHandler<HTMLInputElement> = (e) => {
+    const { name, value } = e.currentTarget;
```
`React.` 사실 암묵적으로 바벨/번들러 레벨에서 해석해주는데... 그래도 `ChangeEventHandler`를 import 하는 건 어떨까요?

taeyoungs replied:
넵넵 import 해오도록 하겠습니다. :)

InSeong-So commented on src/pages/AddingCard/hooks/useOwner.ts:
```diff
@@ -0,0 +1,18 @@
+const useOwner = () => {
+  const [owner, setOwner] = useState('');
+
+  const handleChangeOwner: React.ChangeEventHandler<HTMLInputElement> = (e) => {
+    const { value } = e.target;
+
+    setOwner(value);
+  };
```
이런 코드는 useCallback으로 감싸주면 어떨까요? 😊

InSeong-So reviewed (approved):
안녕하세요, 태영님! 많은 코드가 수정되었는데 너무 좋은 형태로 잘 리팩토링 해주셨어요.
멋있습니다👍👍👍

> useRouter 고도화
- 지금 단계에서는 이것으로 충분합니다! 사실 라우팅 함수가 가장 중요하게 적용되어야하는 것은 **허용되는 경로에 대한 유효성**이거든요. 이게 갖춰지면 나머지는 DX를 위한 영역이기 때문이에요.

> card 접두어
- 컨벤션을 잘 확립해주시고, 코드를 읽으면서 오, 이건 이거겠지? 하면 맞더군요😊 제3자가 이해할 수 있게 일관되게 작성해주시려고 한 점이 너무 좋았어요. 추후, 다른 프로젝트에도 이런 고민을 많이 한 뒤 적용하면 어떨까 싶어요!

> 만료일 관련
- 유저 입장에서는 입력이 되는 것이 맞죠! 하지만 "값을 막는" 행위보다 유저가 인지할 수 있게 에러 메세지를 띄워주는 것은 어떨까요? 폼 제출 시, 또 blur/change/keyup/keypress 시 등이 있겠네요. 이런 폼 핸들링이 제일 복잡하지만, 제어/비제어 컴포넌트 영역에서 많은 고민을 하는 것이 이번 미션의 주된 포인트니까요, 이렇게 옳고 그름을 잘 가려가면서 적용하면 된답니다💪

> addOrUpdateCard 함수 변경
- 코드를 읽으면서 의도가 이해가 되었기 때문에, 좋은 해결책을 찾으신 게 아닐까 싶어요!
- 그래도 id를 만드는 로직을 Date 로 하는 것은 그만두셔도 됩니닷...! `crypto.randomUUID`를 사용해주세요😇

크게 고쳐야 하는 부분은 많이 없어서, 이만 머지하려고 합니다.
리팩토링하느라 정말 고생 많으셨어요!!
👏👏👏
